File: ast/expression/decrement.go
package expression

import (
	"github.com/panda-io/micro-panda/ast"
	"github.com/panda-io/micro-panda/ast/core"
	"github.com/panda-io/micro-panda/ast/types"
)

type Decrement struct {
	ExpressionBase
	Expression ast.Expression
}

func (d *Decrement) Validate(c ast.Context, expected core.Type) {
	d.Expression.Validate(c, expected)
	d.Const = false
	if d.Expression.IsConstant() {
		c.Error(d.GetPosition(), "expect variable")
	}
	if !types.IsInteger(d.Expression.Type()) {
		c.Error(d.GetPosition(), "expect integer expression")
	}
}

File: MicroPanda/AST/Expression/Decrement.cs
namespace MicroPanda.AST.Expression;

using Type;

internal class Decrement : Expression
{
    internal Expression? Expression { get; set; }

    internal override void Validate(Context context, Type? expected)
    {
        _const = false;
        Expression!.Validate(context, expected);
        if (Expression.Const)
        {
            context.Program.Error(Position, "expect variable");
        }
        if (!TypeHelper.IsInteger(Expression.Type!))
        {
            context.Program.Error(Position, "expect integer expression");
        }
    }
}